File: frontend-next/src/types/shareholding.ts
export interface ShareholdingCategory {
  name: string;
  percentage: number;
  shares: number;
}

export interface QuarterlyHolding {
  quarter: string;
  yrc: number;
  promoter: number;
  fii: number;
  dii: number;
  mutual_funds: number;
  insurance: number;
  retail: number;
  others: number;
}

export interface MajorShareholder {
  name: string;
  type: string;
  shares: number;
  percentage: number;
  date: string;
}

export interface ShareholdingResponse {
  symbol: string;
  co_code: number;
  company_name: string;
  sector: string;
  mcap_type: string;
  latest_quarter: string;
  total_shares: number;
  categories: ShareholdingCategory[];
  quarterly_trend: QuarterlyHolding[];
  major_shareholders: MajorShareholder[];
}

// --- Industry-wise trend types ---

export interface IndustryQuarterData {
  quarter: string;
  yrc: number;
  promoter: number;
  fii: number;
  dii: number;
  public: number;
  companies_count: number;
}

export interface IndustryTrendResponse {
  sector: string;
  total_companies: number;
  quarters: IndustryQuarterData[];
}

// --- All-sectors summary types ---

export interface SectorSparklinePoint {
  quarter: string;
  yrc: number;
  value: number;
}

export interface SectorSummaryRow {
  sector: string;
  companies_count: number;
  latest_quarter: string;
  promoter: number;
  fii: number;
  dii: number;
  public: number;
  others: number;
  promoter_trend: SectorSparklinePoint[];
  fii_trend: SectorSparklinePoint[];
  dii_trend: SectorSparklinePoint[];
  public_trend: SectorSparklinePoint[];
  others_trend: SectorSparklinePoint[];
}

export interface AllSectorsSummaryResponse {
  total_sectors: number;
  latest_quarter: string;
  sectors: SectorSummaryRow[];
}

// --- Sector Analytics (cross-database decomposition) types ---

export interface HolderTypeDecomposition {
  holding_value: number;
  prev_holding_value: number | null;
  value_change: number | null;
  price_effect: number | null;
  holding_effect: number | null;
  holding_change_pct: number | null;
  share_pct: number;
}

export interface SectorAggregateMetrics {
  promoter_flow: number | null;
  promoter_change_pct: number | null;
  promoter_accum_index: number | null;
  fii_flow: number | null;
  fii_change_pct: number | null;
  fii_accum_index: number | null;
  dii_flow: number | null;
  dii_change_pct: number | null;
  dii_accum_index: number | null;
  public_flow: number | null;
  public_change_pct: number | null;
  public_accum_index: number | null;
}

export interface SectorQuarterAnalytics {
  quarter: string;
  yrc: number;
  companies_matched: number;
  companies_total: number;
  total_sector_mcap: number;
  promoter: HolderTypeDecomposition;
  fii: HolderTypeDecomposition;
  dii: HolderTypeDecomposition;
  public: HolderTypeDecomposition;
  mcap_weighted: SectorAggregateMetrics;
  equal_weighted: SectorAggregateMetrics;
}

export interface SectorAnalyticsResponse {
  sector: string;
  total_companies: number;
  matched_companies: number;
  unmatched_symbols: string[];
  quarters: SectorQuarterAnalytics[];
}

File: backend/schemas/shareholding.py
from pydantic import BaseModel


class ShareholdingCategory(BaseModel):
    name: str
    percentage: float
    shares: int = 0


class QuarterlyHolding(BaseModel):
    quarter: str        # e.g. "Sep 2025"
    yrc: int            # e.g. 202509
    promoter: float
    fii: float
    dii: float
    mutual_funds: float
    insurance: float
    retail: float
    others: float


class MajorShareholder(BaseModel):
    name: str
    type: str           # e.g. "Promoter", "Public", "FII"
    shares: int = 0
    percentage: float
    date: str = ""


class ShareholdingResponse(BaseModel):
    symbol: str
    co_code: int
    company_name: str
    sector: str = ""
    mcap_type: str = ""
    latest_quarter: str = ""
    total_shares: int = 0
    categories: list[ShareholdingCategory] = []
    quarterly_trend: list[QuarterlyHolding] = []
    major_shareholders: list[MajorShareholder] = []


# --- Industry-wise trend models ---

class IndustryQuarterData(BaseModel):
    quarter: str            # e.g. "Sep 2025"
    yrc: int                # e.g. 202509
    promoter: float
    fii: float
    dii: float
    public: float
    companies_count: int    # how many companies contributed data this quarter


class IndustryTrendResponse(BaseModel):
    sector: str
    total_companies: int
    quarters: list[IndustryQuarterData] = []


class SectorListResponse(BaseModel):
    sectors: list[str]


# --- All-sectors summary models ---

class SectorSparklinePoint(BaseModel):
    quarter: str
    yrc: int
    value: float


class SectorSummaryRow(BaseModel):
    sector: str
    companies_count: int
    latest_quarter: str
    promoter: float
    fii: float
    dii: float
    public: float
    others: float
    promoter_trend: list[SectorSparklinePoint] = []
    fii_trend: list[SectorSparklinePoint] = []
    dii_trend: list[SectorSparklinePoint] = []
    public_trend: list[SectorSparklinePoint] = []
    others_trend: list[SectorSparklinePoint] = []


class AllSectorsSummaryResponse(BaseModel):
    total_sectors: int
    latest_quarter: str
    sectors: list[SectorSummaryRow] = []


# --- Sector Analytics (cross-database decomposition) models ---

class HolderTypeDecomposition(BaseModel):
    holding_value: float
    prev_holding_value: float | None = None
    value_change: float | None = None
    price_effect: float | None = None
    holding_effect: float | None = None
    holding_change_pct: float | None = None
    share_pct: float


class SectorAggregateMetrics(BaseModel):
    promoter_flow: float | None = None
    promoter_change_pct: float | None = None
    promoter_accum_index: float | None = None
    fii_flow: float | None = None
    fii_change_pct: float | None = None
    fii_accum_index: float | None = None
    dii_flow: float | None = None
    dii_change_pct: float | None = None
    dii_accum_index: float | None = None
    public_flow: float | None = None
    public_change_pct: float | None = None
    public_accum_index: float | None = None


class SectorQuarterAnalytics(BaseModel):
    quarter: str
    yrc: int
    companies_matched: int
    companies_total: int
    total_sector_mcap: float
    promoter: HolderTypeDecomposition
    fii: HolderTypeDecomposition
    dii: HolderTypeDecomposition
    public: HolderTypeDecomposition
    mcap_weighted: SectorAggregateMetrics
    equal_weighted: SectorAggregateMetrics


class SectorAnalyticsResponse(BaseModel):
    sector: str
    total_companies: int
    matched_companies: int
    unmatched_symbols: list[str] = []
    quarters: list[SectorQuarterAnalytics] = []
